src/bin/contribute.rs

```diff
@@ -382,66 +386,89 @@ impl Contribute {
+                                    {
+                                        let _ = cloned
+                                            .unlock_chunk(&chunk_id, Some(e.to_string()))
+                                            .await;
+                                    }
+                                    if cloned
+                                        .remove_chunk_id_from_lane_if_exists(
+                                            &PipelineLane::Process,
+                                            &chunk_id,
+                                        )
+                                        .expect("Should have removed chunk ID from lane")
+                                    {
+                                        let _ = cloned
+                                            .unlock_chunk(&chunk_id, Some(e.to_string()))
+                                            .await;
+                                    }
+                                    if cloned
+                                        .remove_chunk_id_from_lane_if_exists(
+                                            &PipelineLane::Upload,
+                                            &chunk_id,
+                                        )
+                                        .expect("Should have removed chunk ID from lane")
+                                    {
+                                        let _ = cloned
+                                            .unlock_chunk(&chunk_id, Some(e.to_string()))
+                                            .await;
+                                    }
+                                    cloned.set_status_update_signal();
                                 }
-                                cloned.set_status_update_signal();
                             }
                         }
+                        tokio::time::delay_for(delay_duration).await;
                     }
-                    tokio::time::delay_for(delay_duration).await;
-                }
-            });
-            futures.push(jh);
-        }
+                });
 
-        futures::future::try_join_all(futures).await?;
+                futures.push(jh);
+            }
 
-        Ok(())
+            match futures::future::try_join_all(futures).await {
+                Err(err) => {
+                    warn!("Crash!!!");
+                    // Afetr crashes have been incremented, one of the threads might have the PIPELINE lock.
+                    // But once that thread releases the lock, we clear the pipeline.
+                    // After that no old thread can add to pipeline lanes.
+                    // If the threads would test before they have the lock, the pipeline might be cleared just before they get the lock.
+                    // Then they could modify the pipeline and cause deadlock.
+                    CRASHES.fetch_add(1, SeqCst);
+                    self.clear_pipeline()?;
```

Comment:
Can we reach a situation here where the following happens?

* A thread gets the write lock and goes over the CRASHES check
* CRASHES gets incremented
* We clear the pipelines
* The thread works on invalid data

If yes, maybe it'd be safer to increment CRASHES while we have the write lock?

Reply:
I think it's possible anyway that a thread will work on invalid data, I'll add something to make sure that the filenames won't conflict.

Reply:
When I wrote "invalid data", I actually meant invalid pipeline data :)

Reply:
Well in any case the property we want to have is that after the pipelines have been cleared, the old threads cannot modify the PIPELINES data structure anymore. I think this holds if we increment the CRASHES counter before or during clearing the pipelines. Only difference should be that in theory more threads can modify the pipeline if we are waiting for the write lock before incrementing the counter.

Here's what should happen in your scenario
* A thread gets the write lock and goes over the CRASHES check
* CRASHES gets incremented
* We clear the pipelines -- clearing the pipelines requires write lock, so the thread has released it before this step is finished
* The thread works on invalid data -- the thread will exit when it tries to modify PIPELINE

Reply:
Got it! That makes sense. Maybe the last thing we should add then is cleanup of older "crash level" files.